### ISSUE TITLE
Feature Reqest: One docker daemon support multiple user namespace

### ISSUE THREAD

**Issue by hanwei1990** [moby/moby#34474]
**Description**

Is it possible to support multiple user namespace in the future? For example,
docker user create testuser
docker run -it --userns testuser ubuntu /bin/bash

Because linux quota command limit disk quota based on uid and gid. 
So we can limit the disk usage of each user while multiple user namespace share one volume.
But now One docker deamon only support to create one user namespace.

Of course, I have a successful attempt on  limiting volume usage for one user namespace.

**Comment by cpuguy83** [comment]
Being able to provide per-container user namespaces is definitely something we want to provide, however due to limitations at the filesystem layer it's just not feasible today.

For every user namespace images have to be chowned to the remapped user/group ID's.
With one userns per daemon, this hit is taken up front when loading a new image.
If we were to provide per-container user namespaces, we'd have to copy the image and chown it specifically for that container (which also negates CoW).

There is work being done on the kernel to support remapping uid/gids of files at the vfs layer on mount, which would unblock this... but this work seems to be a long way from actually getting merged into the kernel, and then beyond that having people actually get this functionality in their distro of choice.
See this LWN article discussing this: https://lwn.net/Articles/700422/
There's also several patchsets if you search for `shiftfs`

**Comment by thaJeztah** [comment]
Yes, this is desirable, but not possible currently (as @cpuguy83 explained). There's an existing issue tracking this, with already some conversation. I'll close this issue in favor of that one, so that the discussion doesn't diverge; https://github.com/moby/moby/issues/28593

**Comment by hanwei1990** [comment]
ok, I will pay close attention on the issue.
